homeassistant/components/bkk/sensor.py

```diff
@@ -0,0 +1,314 @@
+from homeassistant.helpers.entity import Entity
+
+DOMAIN = "bkk"
+CONF_STOP = "stop"
+CONF_GROUPED = "grouped"
+CONF_ONLY_DEPARTURES = "only_departures"
+CONF_MINUTES_AFTER = "minutes_after"
+MAX_DEPARTURES = 5
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_STOP): cv.string,
+        vol.Optional(CONF_GROUPED, default=False): cv.boolean,
+        vol.Optional(CONF_ONLY_DEPARTURES, default=True): cv.boolean,
+        vol.Optional(CONF_MINUTES_AFTER, default=120): cv.positive_int,
+    }
+)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ICONS = {
+    "BUS": "mdi:bus",
+    "TRAM": "mdi:tram-side",
+    "RAIL": "mdi:tram",
+    "SUBWAY": "mdi:subway",
+    "API": "mdi:api",
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_create_group(hass, name, entities):
+    """Create group for bkk route entities."""
+    await group.Group.async_create_group(hass, name, entities, user_defined=True)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up bkk platform."""
+    data_updater = BkkDataUpdater(config)
+    try:
+        data_updater.get_data()
+    except ConnectionError:
+        return
+    else:
+        if data_updater.state == "NOT_FOUND":
+            _LOGGER.error(
+                "Stop id could not be found. Please specify an existing BKK stop id"
+            )
+            return
+        entities = []
+        stop_type = data_updater.data["references"]["stops"][data_updater.stop_id][
+            "type"
+        ]
+        for i in range(MAX_DEPARTURES):
+            route = BkkRoute(data_updater.stop_id, i, stop_type)
+            entities.append(route)
+        data_updater.routes = entities
+        async_add_entities(entities, True)
+        async_add_entities([data_updater], True)
+
+        async def async_remove_unavailable_entities(event):
+            reg = await entity_registry.async_get_registry(hass)
+            entities = reg.entities.copy()
+            for ent in entities:
+                registry_entry = reg.async_get(ent)
+                if registry_entry.platform == DOMAIN:
+                    entity_id = registry_entry.entity_id
+                    state = hass.states.get(entity_id).state
+                    if state == STATE_UNAVAILABLE:
+                        reg.async_remove(entity_id)
+
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, async_remove_unavailable_entities
+        )
+        return True
+
+
+class BkkDataUpdater(Entity):
+    """Data updater class to query api and update entity data."""
+
+    def __init__(self, config):
+        """Initialize the sensor."""
+        self._name = "BKK Data Updater"
+        self.api = futar.BkkFutar()
+        self.routes = None
+        self.stop_id = (
+            f"BKK_{config[CONF_STOP][1:]}"
+            if config[CONF_STOP][0] == "#"
+            else f"BKK_{config[CONF_STOP]}"
+        )
```

Comment:
Please make this more readable by expanding the if else statements.

Reply:
fixed